### ISSUE TITLE
info.json error in cache

### ISSUE THREAD

**Issue by devanonyme** [medusa-project/cantaloupe#84]
Hello,

My coworker have installed the new version of Cantaloupe 3.2.2, thanks for the update! We will retest it with the issues we had encountered! We have released our first implementation of Cantaloupe and Openseadragon in production this week and we will add more higher resolution content (maps and plans) next year.

While testing I have encountered a bug I had noticed previously, I had mentionned it in another bug, so I open it here with an exemple (see the attached image).

Yesterday I had accessed the new server while it was missing some configs and the json was returned in error. I have reloaded the page today, with the server up to date, and the same json in error was returned from cache :

![error_in_cache](https://cloud.githubusercontent.com/assets/21067311/21393271/b9d91b86-c761-11e6-96f8-1da6114877ca.PNG)

My intuition is that there should not be cache header for request returning an error but I'm not sure if the header are coming from Cantaloupe or the proxy.

I will double check that with my coworker but I report the issue here in case you would have already encountered this issue.

**Comment by devanonyme** [comment]
I had opened a related issue OpenSeadragon side (https://github.com/openseadragon/openseadragon/issues/1086) but that would also cover the case of momentary lost of network.

At the simplest, I can get over the cache error with this :

    var _open_failed_retry = function (event) {
        var retry_open_tile_source = event.source + "?bustcache=" + (new Date()).getTime();
        _viewerInstance.open(retry_open_tile_source);
    };

   _viewerInstance.addHandler("open-failed", _open_failed_retry);

I will check the case of network lost and improve on this (number of attemps + delay by exemple).

**Comment by adolski** [comment]
Hi @devanonyme,

The `Cache-Control` header is intended to work as described below, and I have confirmed that it does, in version 3.2.2:

#### `cache.client.enabled = true`
* 2xx: `Cache-Control` header returned according to `cache.client.*` keys in configuration
* 3xx: No `Cache-Control` header
* 4xx & 5xx: `Cache-Control: must-revalidate,no-cache,no-store`

#### `cache.client.enabled = false`
* 2xx & 3xx: No `Cache-Control` header
* 4xx & 5xx: `Cache-Control: must-revalidate,no-cache,no-store` (Restlet adds this automatically, I haven't looked into how to disable it, but it doesn't really matter I think)

So, I would suspect the reverse proxy. Feel free to reopen this issue if that is not the case.

Also, there is nothing special about the `cache.client.*` keys; they just add a `Cache-Control` header to successful requests. If it makes things easier for you, you could turn off this header in Cantaloupe and configure your reverse proxy to add it instead.

P.S. I'm happy that Cantaloupe is working out for you!

**Comment by devanonyme** [comment]
Hello,
 
I'm not sure you will see this or if I should open a new bug. I was able to reproduce the issue with 3.3 standalone.

For some reasons we sometimes have url to images before the images are really available through the filesystem (the resolver will return a valid path but not to an existing ressource) or we also have a bug where sometimes our dev environnement reboot and the mount lose it's access right.

I had noticed in such circumstances that the json file (404 error) was then put in browser cache and when the file was present the json file in error was still used (I was able to partly solve the problem by implementing a bust cache / retry n times in OpenSeadragon on failed access).

To reproduce :
1) I use Cantaloupe 3.3 standalone
2) Access a non existing ressource http://localhost:8182/iiif/2/image2.jpg/info.json through our web application (I was not able to reproduce it when entering the url in standalone in browser adress?)

This is what I see while inspecting the logs (notice the 404 from disk cache)

![screen shot 05-04-17 at 11 31 am](https://cloud.githubusercontent.com/assets/21067311/25712146/d03de642-30be-11e7-91b0-18d272eca0a6.PNG)

I will investigate some more but this is what I have currently....

**Comment by devanonyme** [comment]
It seem to be a sequence issue.

Cache directives of "informationRessource.java" is added too soon, before the response is known to be in error.

It's probably not the right way to fix it but I have moved "getResponseCacheDirectives().addAll(getCacheDirectives());" at the end of "Representation doGet()"

```
        getResponseCacheDirectives().addAll(getCacheDirectives());
        return rep;
```
I then get the appropriate behavior (caching on good answer, no cache on error (at least the 404 from my test)).

**Comment by adolski** [comment]
I'm not sure that moving that line has any effect. When an exception is thrown in `doGet()` (including a FileNotFoundException from a resolver), [e.i.l.c.WebApplication$CustomStatusService](https://restlet.com/open-source/documentation/javadocs/2.3/jse/api/org/restlet/service/StatusService.html) takes over and returns its own headers.

What's interesting, though, is the warning in your debugger: "Provisional headers are shown". I didn't know what that meant, but [this person](http://stackoverflow.com/a/21179105) explains it.

In any case, when I invoke `curl -I http://localhost:8182/iiif/2/missing.jpg/info.json` the response includes a `Cache-Control: must-revalidate,no-cache,no-store` header, no matter whether the directives are added in `doInit()` or `doGet()`. If you can figure out what headers Chrome is sending, then Cantaloupe should respond the same way when the same headers are sent via curl, and then I would be able to reproduce this. But from the looks of it, Chrome might not even be sending a request at all (?).

**Comment by devanonyme** [comment]
Hello, I guess you where in vacation, so welcome back!

I'm not sure what caused "Provisional headers are show" but I made sure to disable all my extensions (Browserstack, Netbeans and Postman)  and did the test again, here is the result with original code :

![screen shot 05-09-17 at 03 08 pm](https://cloud.githubusercontent.com/assets/21067311/25868067/9ee63540-34c9-11e7-9b60-b60ed5b23512.PNG)

With the hack (moved line) 
![screen shot 05-09-17 at 03 12 pm](https://cloud.githubusercontent.com/assets/21067311/25868240/021d1a5c-34ca-11e7-8d6b-51d0002a5506.PNG)

I will check Chromes headers and also test with curl to validate if I see the same thing as you.

**Comment by devanonyme** [comment]
There is indeed no difference in the result when using curl (i did the test from a remote linux to my windows machine)

With orignal source :
curl -I http://10.9.1.58:8182/iiif/2/missing.jpg/info.json
HTTP/1.1 404 Not Found
Date: Tue, 09 May 2017 19:19:40 GMT
Date: Tue, 09 May 2017 19:19:42 GMT
Accept-Ranges: bytes
Server: Restlet-Framework/2.3.8
Vary: Accept-Charset, Accept-Encoding, Accept-Language, Accept
X-Powered-By: Cantaloupe/Unknown
Content-Type: text/html;charset=iso-8859-1
Cache-Control: must-revalidate,no-cache,no-store
Content-Length: 345


With the moved line :
curl -I http://10.9.1.58:8182/iiif/2/missing.jpg/info.json
HTTP/1.1 404 Not Found
Date: Tue, 09 May 2017 19:18:20 GMT
Date: Tue, 09 May 2017 19:18:20 GMT
Accept-Ranges: bytes
Server: Restlet-Framework/2.3.8
Vary: Accept-Charset, Accept-Encoding, Accept-Language, Accept
X-Powered-By: Cantaloupe/Unknown
Content-Type: text/html;charset=iso-8859-1
Cache-Control: must-revalidate,no-cache,no-store
Content-Length: 345

**Comment by devanonyme** [comment]
This is the request headers that I can see in debug when doing the request from Chrome (unmodified code) 
![screen shot 05-09-17 at 03 29 pm](https://cloud.githubusercontent.com/assets/21067311/25868923/6b44b4c0-34cc-11e7-9c12-54f9edb667bb.PNG)

**Comment by devanonyme** [comment]
Maybe it's cache-control that changes the behavior?

**Comment by devanonyme** [comment]
With all the same props
curl -I http://10.9.1.58:8182/iiif/2/missing2.jpg/info.json -H 'Accept:text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8' -H 'Upgrade-Insecure-Requests:1' -H 'Connection:keep-alive' -H 'User-Agent:Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36' -H 'Cache-Control: max-age=0' -H 'Accept-Encoding:gzip, deflate, sdch, br' -H 'Accept-Language:en-US;en;q=0.8'
HTTP/1.1 404 Not Found
Date: Tue, 09 May 2017 20:20:42 GMT
Date: Tue, 09 May 2017 20:21:24 GMT
Accept-Ranges: bytes
Server: Restlet-Framework/2.3.8
Vary: Accept-Charset, Accept-Encoding, Accept-Language, Accept
X-Powered-By: Cantaloupe/Unknown
Content-Type: text/html;charset=iso-8859-1
Cache-Control: must-revalidate,no-cache,no-store
Content-Length: 346

**Comment by adolski** [comment]
Looks like strange Chrome behavior, no?

**Comment by devanonyme** [comment]
Yes!

It's not specific to Chrome, I was able to reproduce it in Firefox and Explorer.
I was wondering if there could be a lower level mecanism (in restlet / jetty) that could prevent the manipulation of user agent by exemple (why manipulating the headers would not help reproduce the issue with curl).

I will try to see if I can think of something else to check, let met know if there is anything you would like me to try out on my side

**Comment by devanonyme** [comment]
Can you reproduce it with a Browser or the issue is only on my side?

**Comment by devanonyme** [comment]
Maybe the cache-control is overriden if only the head is asked? (ie. -I option)? That would make sense as caching the request without the body would be bad...

**Comment by devanonyme** [comment]
I think that's the problem, here is the result with -i instead i found in curl documentation (For HTTP, you can get the header information (the same as -I would show) shown before the data by using -i/--include.)

curl http://10.9.1.58:8182/iiif/2/missing2.jpg/info.json -i -H 'Accept:text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8' -H 'Upgrade-Insecure-Requests:1' -H 'Connection:keep-alive' -H 'User-Agent:Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36' -H 'Accept-Encoding:gzip, deflate, sdch, br' -H 'Accept-Language:en-US;en;q=0.8'
HTTP/1.1 404 Not Found
Date: Wed, 10 May 2017 15:04:57 GMT
Content-Type: text/html;charset=utf-8
Cache-Control: max-age=2592000, public, no-transform
Date: Wed, 10 May 2017 15:04:57 GMT
Accept-Ranges: bytes
Server: Restlet-Framework/2.3.8
Vary: Accept-Charset, Accept-Encoding, Accept-Language, Accept
X-Powered-By: Cantaloupe/Unknown
Transfer-Encoding: chunked

<!DOCTYPE html><html>    <head>            <base href="http://10.9.1.58:8182/">
<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">

<link rel="stylesheet" href="static/styles/bootstrap.min.css">
        <title>404 Not Found</title>    </head>    <body id="cl-error">        <div class="container-fluid">            <h1>404 Not Found</h1>

**Comment by adolski** [comment]
`HEAD` vs. `GET` was my problem. I can reproduce it now 😄

**Comment by adolski** [comment]
This is now fixed on the `release/3.3` branch, and it has been merged into `develop`.

Thanks for all of your investigation, @devanonyme!

**Comment by devanonyme** [comment]
Thanks to you!
